### ISSUE TITLE
SDXL "load_lora_weights" results in different ouputs than loading in unet and text encoder in two steps

### ISSUE THREAD

**Issue by To-jak** [huggingface/diffusers#6639]
### Describe the bug

I seem to get very different results depending on how I load the weights of my trained SDXL LoRA.

* In the first experience, I load the weights with the `load_lora_weights` method.

Here is the output for basic text to image pipeline inference:
![first_method_resized_256](https://github.com/huggingface/diffusers/assets/33063955/cd279919-1abc-44a9-97dd-e3d19870585d)

* In the second one, I load the weights with `load_lora_into_unet` and `load_lora_into_text_encoder`.

The output is different (and looks much better imo):
![second_method_resized_256](https://github.com/huggingface/diffusers/assets/33063955/c8279e75-136f-42d0-8fa1-3d5c9e93aff5)

The same generator seed is used for inference.

I got my lora model weights using the [train_dreambooth_lora_sdxl_advanced.py](https://github.com/huggingface/diffusers/blob/v0.25.1/examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py) example script and the [dog images](https://huggingface.co/datasets/diffusers/dog-example) from documentation.

The training script performs some operations to convert the state_dict to diffusers, but it does so by using the `diffusers.utils` methods. Anyway, I find it odd to observe this behavior. Thanks! 

### Reproduction

**First experience**, loading the lora weights with the `load_lora_weights` method:

```Python
import torch
from diffusers import StableDiffusionXLPipeline

pipe = StableDiffusionXLPipeline.from_pretrained(
    'stabilityai/stable-diffusion-xl-base-1.0',
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True
)
pipe.to('cuda')

# Load LoRA with "load_lora_weights"
pipe.load_lora_weights(
    # Path to my trained lora output_dir
    '.cache/training_artifacts/626d7ad2-c5e8-4d58-88a9-42a3114f852d',
    weight_name="pytorch_lora_weights.safetensors",
    adapter_name='test_lora'
)
pipe.set_adapters(["test_lora"], adapter_weights=[1.0])

prompt = 'picture of ohwx character in a bucket'

generator = torch.Generator(device="cuda").manual_seed(42)
pipe(
    prompt=prompt,
    num_inference_steps=20,
    generator=generator
).images[0]
```

**Second experience**, loading the lora weights in two steps, unet then text encoder:

```python
import torch
from diffusers import StableDiffusionXLPipeline

pipe = StableDiffusionXLPipeline.from_pretrained(
    'stabilityai/stable-diffusion-xl-base-1.0',
    torch_dtype=torch.float16,
    variant="fp16",
    use_safetensors=True
)
pipe.to('cuda')

# Load LoRA with "load_lora_into_unet" and "load_lora_into_text_encoder"
state_dict, network_alphas = StableDiffusionXLPipeline.lora_state_dict(
    # Path to my trained lora output_dir
    '.cache/training_artifacts/626d7ad2-c5e8-4d58-88a9-42a3114f852d',
    weight_name="pytorch_lora_weights.safetensors"
)
pipe.load_lora_into_unet(state_dict, network_alphas, pipe.unet, adapter_name='test_lora')
pipe.load_lora_into_text_encoder(state_dict, network_alphas, pipe.text_encoder, adapter_name='test_lora')
pipe.set_adapters(["test_lora"], adapter_weights=[1.0])

prompt = 'picture of ohwx character in a bucket'

generator = torch.Generator(device="cuda").manual_seed(42)
pipe(
    prompt=prompt,
    num_inference_steps=20,
    generator=generator
).images[0]

```

### System Info

Python: `3.10.12`
diffusers: `0.25.0`
peft: `0.7.1`
accelerate: `0.23.0`
transformers: `4.36.0`
Using GPU in script?: Yes
advanced training script commit hash: https://github.com/huggingface/diffusers/commit/7f551e29ff4ad05615cb38530a8940811f9e5936

### Who can help?

@sayakpaul

**Comment by sayakpaul** [comment]
Cc: @linoytsaban for advanced trainer. 

But in any case we don't have access tp ".cache/training_artifacts/626d7ad2-c5e8-4d58-88a9-42a3114f852d".

**Comment by linoytsaban** [comment]
Hey @To-jak!
Did you use pivotal tuning to train your LoRA? - i.e. enable `--train_text_encoder_ti`?

**Comment by To-jak** [comment]
I made the model weights available here: https://huggingface.co/Tojak/test_lora-dog-626d7ad2-c5e8 🙂

Here are some additional info about the training parameters (did not use `train_text_encoder_ti`):

```yaml
training_parameters:
  adam_beta2: 0.99
  adam_weight_decay: 0.01
  checkpointing_steps: 10000
  instance_prompt: a ohwx character
  learning_rate: 1.0
  max_train_steps: 300
  optimizer: prodigy
  pretrained_model_name_or_path: stabilityai/stable-diffusion-xl-base-1.0
  pretrained_vae_model_name_or_path: madebyollin/sdxl-vae-fp16-fix
  rank: 32
  repeats: 1
  resolution: 1024
  seed: 42
  snr_gamma: 5.0
  text_encoder_lr: 1.0
  train_batch_size: 2
  train_text_encoder: true
  train_text_encoder_frac: 0.8
  train_text_encoder_ti: false
```

**Comment by linoytsaban** [comment]
Thanks! did you experience this also with LoRAs trained with the regular- [train_dreambooth_lora_sdxl.py] (https://github.com/huggingface/diffusers/blob/v0.25.1/examples/dreambooth/train_dreambooth_lora_sdxl.py) script? 

The main difference between the two scripts is the pivotal tuning feature (re: `--train_text_encoder_ti`) so it might help us pinpoint the issue.

**Comment by asomoza** [comment]
that doesn't seem to be the issue here, I just loaded the lora locally and in a collab with the exact same code and it works fine, this seems to be an error in @To-jak's end, maybe you used a different lora model for the first one? it seems like a difference in the training steps, since the first one doesn't even have a bucket.

See here: https://colab.research.google.com/drive/1rUX_gvUp8E-6fBJjJQrOxRPjv9iO4bLr?usp=sharing

<img src="https://github.com/huggingface/diffusers/assets/5442875/f008f87e-4fc5-4128-be1a-6a2de6d4cb0a" width="512" height="512">

**Comment by To-jak** [comment]
I made of copy of your notebook @asomoza, which allows me to highlight the problem better (thank you!)

See here: https://colab.research.google.com/drive/1niM8iAPDt3DQN3GSUPO9HTmXhNw03P9C?usp=drive_link

It's actually the same lora model (hence the fact that there is only one bucket).
You can see that the result images look different depending on how I load the model:

First method:
![first_method](https://github.com/huggingface/diffusers/assets/33063955/2fbb066a-5147-4c56-9b31-b3eeca392ab0)

Second method:
![second_method](https://github.com/huggingface/diffusers/assets/33063955/cf183392-3b7d-4eac-bedb-105446c8c3e6)

@linoytsaban I don't know if the problem persists with the regular training script, I'm going to perform a training to investigate on my side.

**Comment by asomoza** [comment]
oh I see, is that the first image was anything like these ones, I though your problem was that the load_lora_weights was producing pictures like the white dog one. 

SDXL has two text encoders, the load_lora_weights loads the lora in both of them and you are just loading it in the first one with the second method, could you try loading it also in the second text encoder and see if the results are the same?

**Comment by asomoza** [comment]
@To-jak I had some spare time and did it myself, indeed that is the difference, I updated the colab for you to see the results:

https://colab.research.google.com/drive/1rUX_gvUp8E-6fBJjJQrOxRPjv9iO4bLr?usp=sharing

I changed the generator to "cpu" so the result are exact in the three generations.

If what you're asking if the results are better or worse with one text encoder or both of them, I think I saw in the kohya repo that training both text encoders could make it worse but for me that isn't the case. For example if I use your lora with my usual settings and loaded in both text encoders I get this image:

![20240119113428](https://github.com/huggingface/diffusers/assets/5442875/a3841993-81b6-49da-ac76-ccea7601dfff)

**Comment by To-jak** [comment]
@asomoza Yes you are right, when I load the LoRA in the two text encoders, I got the same result as with the `load_lora_weights` method.

```python
# Load LoRA in unet, and in the two text encoders
state_dict, network_alphas = StableDiffusionXLPipeline.lora_state_dict(
    'Tojak/test_lora-dog-626d7ad2-c5e8',
    weight_name="pytorch_lora_weights.safetensors"
)
pipe.load_lora_into_unet(state_dict, network_alphas, pipe.unet, adapter_name='test_lora')
pipe.load_lora_into_text_encoder(state_dict, network_alphas, pipe.text_encoder, prefix='text_encoder', adapter_name='test_lora')
pipe.load_lora_into_text_encoder(state_dict, network_alphas, pipe.text_encoder_2, prefix='text_encoder_2', adapter_name='test_lora')
pipe.set_adapters(["test_lora"], adapter_weights=[1.0])
```

<img src="https://github.com/huggingface/diffusers/assets/33063955/9afd7e25-052f-48d6-9735-5def62a7bd6e" width="256" height="256">


Actually you helped me pinpoint that something is wrong on my end with my environment, as I get a completely different result (with both ways of loading the LoRA model):

<img src="https://github.com/huggingface/diffusers/assets/33063955/81f5a41e-3998-4909-8428-c98bd427b0c6" width="256" height="256">

That was the first image I posted. I guess I need to investigate further into my local env setup.

**Comment by To-jak** [comment]
Some news after more investigation on my part: I was able to pinpoint that the degradation of my results in my environment seems to be coming from the way I get the seed `generator`.

Using `torch.Generator(device="cuda").manual_seed(42)`, the model output looks degraded and differs from what I obtain in other environments like Colab for example. By using `torch.manual_seed(42)`, I obtain a result consistent with what you were able to get, and that doesn't seem degraded to me.

**Comment by yiyixuxu** [comment]
I think we can close this issue now? :)
thanks again @asomoza !

**Comment by To-jak** [comment]
Yes, at the end the problem I was facing was not related to the issue title anyway :)
Thanks!
